base_name_search_improved/models/ir_model.py

```diff
@@ -0,0 +1,76 @@
+# -*- coding: utf-8 -*-
+# © 2016 Daniel Reis
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from openerp import models, fields, api
+from openerp import SUPERUSER_ID
+from openerp import tools
+
+
+# Extended name search is only used on some operators
+ALLOWED_OPS = set(['ilike', 'like'])
+
+
+@tools.ormcache(skiparg=0)
+def _get_rec_names(self):
+    model = self.env['ir.model'].search(
+        [('model', '=', str(self._model))])
+    rec_name = [self._rec_name] or []
+    other_names = model.name_search_ids.mapped('name')
+    return rec_name + other_names
+
+
+def _extend_name_results(self, domain, results, limit):
+    result_count = len(results)
+    if result_count < limit:
+        domain += [('id', 'not in', [x[0] for x in results])]
+        recs = self.search(domain, limit=limit - result_count)
+        results.extend(recs.name_get())
+    return results
```

Comment:
These is just helper code, wrapped in function for convenience.
They are not `ir.model` methods, since the `self ` parameter passed to them is the searched Model, and not `ir,.model`.